src/mlpack/methods/ann/layer/maxout_impl.hpp

```diff
@@ -0,0 +1,134 @@
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_MAXOUT_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_MAXOUT_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "maxout.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+Maxout<InputDataType, OutputDataType>::Maxout()
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+Maxout<InputDataType, OutputDataType>::Maxout(
+    const size_t inSize,
+    const size_t hiddenSize,
+    const size_t outSize) :
+    inSize(inSize),
+    hiddenSize(hiddenSize),
+    outSize(outSize)
+{
+  weights.set_size((inSize * hiddenSize * outSize) + (outSize * hiddenSize), 1);
+}
+
+template<typename InputDataType, typename OutputDataType>
+void Maxout<InputDataType, OutputDataType>::Reset()
+{
+  weight = arma::mat(weights.memptr(), outSize * hiddenSize, inSize,
+      false, false);
+  bias = arma::mat(weights.memptr() + weight.n_elem,
+      outSize * hiddenSize, 1, false, false);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Maxout<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (output.n_elem == 0)
+    output = arma::zeros<arma::mat>(outSize, input.n_cols);
+
+  hiddenParameter = (weight * input) + bias;
+
+  for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++)
+    output.row(j) = arma::max(hiddenParameter.rows(i, i + hiddenSize - 1), 0);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Maxout<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>&& /* input */, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
+{
+  arma::mat hiddenGradient = arma::zeros<arma::mat>(hiddenParameter.n_rows,
+      hiddenParameter.n_cols);
+  arma::mat hiddenParameterSubset;
+
+  for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++)
+  {
+    hiddenParameterSubset = hiddenParameter.rows(i, i + hiddenSize - 1);
+    
+    for (int k = 0; k < hiddenParameterSubset.n_cols; k++)
+    {
+      int ind = arma::as_scalar(arma::find(arma::max(
+          hiddenParameterSubset.col(k)) == hiddenParameterSubset.col(k)));
+      hiddenGradient(i + ind, k) = gy(j, k);
+    }
+  }
+
+  g = weight.t() * hiddenGradient;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void Maxout<InputDataType, OutputDataType>::Gradient(
+    const arma::Mat<eT>&& input,
+    arma::Mat<eT>&& error,
+    arma::Mat<eT>&& gradient)
+{
+  arma::mat hiddenGradient = arma::zeros<arma::mat>(hiddenParameter.n_rows,
+      hiddenParameter.n_cols);
+  arma::mat hiddenParameterSubset;
+
+  for (int i = 0, j = 0; j < outSize; i += hiddenSize, j++)
+  {
+    hiddenParameterSubset = hiddenParameter.rows(i, i + hiddenSize - 1);
+    
+    for (int k = 0; k < hiddenParameterSubset.n_cols; k++)
+    {
+      int ind = arma::as_scalar(arma::find(arma::max(
```

Comment:
It looks right, it's basically what you do for MaxPooling, where I also used the same approach:

```
class MaxPoolingRule
{
 public:
  /*
   * Return the maximum value within the receptive block.
   *
   * @param input Input used to perform the pooling operation.
   */
  template<typename MatType>
  size_t Pooling(const MatType& input)
  {
    return arma::as_scalar(arma::find(input.max() == input, 1));
  }
};
```

But I think I can construct a case where max isn't just a single element e.g. ``arma::max({0, 0, 0})`` or ``arma::max({0.5, 0.5, 0})`` here max returns more than one element, which isn't a problem, we can just use the first max. Besides, specifying the number would probably result in a speedup since you can break once you found the first occurrences of the maximum. Let me know what you think.